// @chakra-ui/layout/dist/cjs/stack.js
"use strict";

exports.__esModule = true;
exports.VStack = exports.HStack = exports.Stack = exports.StackItem = exports.StackDivider = void 0;

var _system = require("@chakra-ui/system");

var _utils = require("@chakra-ui/utils");

var React = _interopRequireWildcard(require("react"));

function _getRequireWildcardCache() { if (typeof WeakMap !== "function") return null; var cache = new WeakMap(); _getRequireWildcardCache = function _getRequireWildcardCache() { return cache; }; return cache; }

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } if (obj === null || typeof obj !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _objectWithoutPropertiesLoose(source, excluded) { if (source == null) return {}; var target = {}; var sourceKeys = Object.keys(source); var key, i; for (i = 0; i < sourceKeys.length; i++) { key = sourceKeys[i]; if (excluded.indexOf(key) >= 0) continue; target[key] = source[key]; } return target; }

function _extends() { _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; }; return _extends.apply(this, arguments); }

var StackDivider = function StackDivider(props) {
  return /*#__PURE__*/React.createElement(_system.chakra.div, _extends({
    className: "chakra-stack__divider"
  }, props, {
    __css: _extends({}, props.__css, {
      borderWidth: 0,
      alignSelf: "stretch",
      borderColor: "inherit",
      width: "auto",
      height: "auto"
    })
  }));
};

exports.StackDivider = StackDivider;

var StackItem = function StackItem(props) {
  return /*#__PURE__*/React.createElement(_system.chakra.div, _extends({
    className: "chakra-stack__item"
  }, props, {
    __css: _extends({
      display: "inline-block",
      flex: "0 0 auto",
      minWidth: 0
    }, props["__css"])
  }));
};

exports.StackItem = StackItem;

/**
 * If we ever run into SSR issues with this, check this post to find a fix for it:
 * @see https://medium.com/@emmenko/patching-lobotomized-owl-selector-for-emotion-ssr-5a582a3c424c
 */
var selector = "& > *:not(style) ~ *:not(style)";
/**
 * Stacks help you easily create flexible and automatically distributed layouts
 *
 * You can stack elements in the horizontal or vertical direction,
 * and apply a space or/and divider between each element.
 *
 * It uses `display: flex` internally and renders a `div`.
 *
 * @see Docs https://chakra-ui.com/components/stack
 *
 */

var Stack = /*#__PURE__*/(0, _system.forwardRef)(function Stack(props, ref) {
  var _ref2;

  var isInline = props.isInline,
      directionProp = props.direction,
      align = props.align,
      justify = props.justify,
      _props$spacing = props.spacing,
      spacing = _props$spacing === void 0 ? "0.5rem" : _props$spacing,
      wrap = props.wrap,
      children = props.children,
      divider = props.divider,
      className = props.className,
      shouldWrapChildren = props.shouldWrapChildren,
      rest = _objectWithoutPropertiesLoose(props, ["isInline", "direction", "align", "justify", "spacing", "wrap", "children", "divider", "className", "shouldWrapChildren"]);

  var direction = isInline ? "row" : directionProp != null ? directionProp : "column";
  var styles = React.useMemo(function () {
    var _ref;

    var directionStyles = {
      column: {
        mt: spacing,
        ml: 0
      },
      row: {
        ml: spacing,
        mt: 0
      },
      "column-reverse": {
        mb: spacing,
        mr: 0
      },
      "row-reverse": {
        mr: spacing,
        mb: 0
      }
    };
    return _ref = {
      flexDirection: direction
    }, _ref[selector] = (0, _utils.mapResponsive)(direction, function (value) {
      return directionStyles[value];
    }), _ref;
  }, [direction, spacing]);
  /**
   * Divider Styles
   */

  var marginX = (0, _utils.mapResponsive)(direction, function (value) {
    return value.includes("row") ? spacing : 0;
  });
  var marginY = (0, _utils.mapResponsive)(direction, function (value) {
    return value.includes("row") ? 0 : spacing;
  });
  var borderLeftWidth = (0, _utils.mapResponsive)(direction, function (value) {
    return value.includes("row") ? "1px" : 0;
  });
  var borderBottomWidth = (0, _utils.mapResponsive)(direction, function (value) {
    return value.includes("row") ? 0 : "1px";
  });
  var dividerStyles = {
    marginX: marginX,
    marginY: marginY,
    borderLeftWidth: borderLeftWidth,
    borderBottomWidth: borderBottomWidth
  };
  var hasDivider = !!divider;
  var shouldUseChildren = !shouldWrapChildren && !hasDivider;
  var validChildren = (0, _utils.getValidChildren)(children);
  var clones = shouldUseChildren ? validChildren : validChildren.map(function (child, index) {
    var isLast = index + 1 === validChildren.length;
    var wrappedChild = /*#__PURE__*/React.createElement(StackItem, {
      key: index
    }, child);

    var _child = shouldWrapChildren ? wrappedChild : child;

    if (!hasDivider) return _child;
    var clonedDivider = /*#__PURE__*/React.cloneElement(divider, dividerStyles);

    var _divider = isLast ? null : clonedDivider;

    return /*#__PURE__*/React.createElement(React.Fragment, {
      key: index
    }, [_child, _divider]);
  });

  var _className = (0, _utils.cx)("chakra-stack", className);

  return /*#__PURE__*/React.createElement(_system.chakra.div, _extends({
    ref: ref,
    display: "flex",
    alignItems: align,
    justifyContent: justify,
    flexDirection: styles.flexDirection,
    flexWrap: wrap,
    className: _className,
    __css: hasDivider ? {} : (_ref2 = {}, _ref2[selector] = styles[selector], _ref2)
  }, rest), clones);
});
exports.Stack = Stack;

if (_utils.__DEV__) {
  Stack.displayName = "Stack";
}
/**
 * A view that arranges its children in a horizontal line.
 */


var HStack = /*#__PURE__*/(0, _system.forwardRef)(function (props, ref) {
  return /*#__PURE__*/React.createElement(Stack, _extends({
    align: "center"
  }, props, {
    direction: "row",
    ref: ref
  }));
});
exports.HStack = HStack;

if (_utils.__DEV__) {
  HStack.displayName = "HStack";
}
/**
 * A view that arranges its children in a vertical line.
 */


var VStack = /*#__PURE__*/(0, _system.forwardRef)(function (props, ref) {
  return /*#__PURE__*/React.createElement(Stack, _extends({
    align: "center"
  }, props, {
    direction: "column",
    ref: ref
  }));
});
exports.VStack = VStack;

if (_utils.__DEV__) {
  VStack.displayName = "VStack";
}
